File: internal/core/domain/graph.go
package domain

import "time"

// GraphGroup defines the category of a node.
type GraphGroup string

const (
	GroupAP      GraphGroup = "ap"
	GroupStation GraphGroup = "station"
	GroupNetwork GraphGroup = "network"
)

// GraphNode represents a node in the visualization graph.
// Modularized using composition to separate concerns while maintaining the legacy JSON API.
// Anonymous embedding ensures field promotion for both Go code and JSON serialization.
type GraphNode struct {
	NodeIdentity
	RadioDetails
	TrafficStats
	ConnectionDetails

	// State and Metadata
	IsStale bool `json:"is_stale,omitempty"`
}

// NodeIdentity encapsulates basic identification and classification.
type NodeIdentity struct {
	ID        string     `json:"id"`
	Label     string     `json:"label"`
	Group     GraphGroup `json:"group"` // "ap", "station", "network"
	MAC       string     `json:"mac,omitempty"`
	Vendor    string     `json:"vendor,omitempty"`
	Signature string     `json:"signature,omitempty"`
	Model     string     `json:"model,omitempty"`
	OS        string     `json:"os,omitempty"`
	FirstSeen time.Time  `json:"first_seen,omitempty"`
	LastSeen  time.Time  `json:"last_seen,omitempty"`
}

// RadioDetails encapsulates WiFi physical and link layer attributes.
type RadioDetails struct {
	// Strings and Slices (16-24 bytes)
	SSID          string   `json:"ssid,omitempty"`
	Security      string   `json:"security,omitempty"`
	Standard      string   `json:"standard,omitempty"`
	WPSInfo       string   `json:"wps_info,omitempty"` // "Configured", "Unconfigured" or empty
	HandshakeFile string   `json:"handshake_file,omitempty"`
	Capabilities  []string `json:"capabilities,omitempty"`
	ProbedSSIDs   []string `json:"probedSSIDs,omitempty"`
	IETags        []int    `json:"ieTags,omitempty"`

	// Pointers (8 bytes)
	RSNInfo        *RSNInfo        `json:"rsn_info,omitempty"`
	WPSDetails     *WPSDetails     `json:"wps_details,omitempty"`
	MobilityDomain *MobilityDomain `json:"mobility_domain,omitempty"`

	// Numerics (8 bytes)
	Channel      int `json:"channel,omitempty"`
	ChannelWidth int `json:"bw"`
	Frequency    int `json:"frequency,omitempty"`
	RSSI         int `json:"rssi,omitempty"`

	// Booleans (1 byte)
	IsWiFi6      bool `json:"is_wifi6,omitempty"`
	IsWiFi7      bool `json:"is_wifi7,omitempty"`
	IsRandomized bool `json:"is_randomized,omitempty"`
	HasHandshake bool `json:"has_handshake,omitempty"`
}

// TrafficStats captures data transmission metrics.
type TrafficStats struct {
	DataTransmitted int64 `json:"data_tx"`
	DataReceived    int64 `json:"data_rx"`
	PacketsCount    int   `json:"packets"`
	RetryCount      int   `json:"retries"`
}


// ConnectionDetails captures current association state.
type ConnectionDetails struct {
	ConnectionState  ConnectionState `json:"connection_state,omitempty"`
	ConnectionTarget string          `json:"connection_target,omitempty"`
	ConnectionError  string          `json:"connection_error,omitempty"`
}

// EdgeType defines the nature of the connection between nodes.
type EdgeType string

const (
	TypeConnection  EdgeType = "connection"
	TypeProbe       EdgeType = "probe"
	TypeCorrelation EdgeType = "correlation"
)

// GraphEdge represents a connection between two nodes.
type GraphEdge struct {
	From   string   `json:"from"`
	To     string   `json:"to"`
	Dashed bool     `json:"dashed,omitempty"`
	Type   EdgeType `json:"type,omitempty"`
	Label  string   `json:"label,omitempty"` // For display
	Color  string   `json:"color,omitempty"` // Hex or rgba for dynamic override
}

// GraphData allows sending the whole graph state to the frontend.
type GraphData struct {
	Nodes  []GraphNode  `json:"nodes"`
	Edges  []GraphEdge  `json:"edges"`
	Config *GraphConfig `json:"config,omitempty"`
}

type GraphConfig struct {
	RSSIThresholds RSSIThresholds `json:"rssi_thresholds"`
	Colors         GraphColors    `json:"colors"`
}

type RSSIThresholds struct {
	Good int `json:"good"` // e.g. -65
	Fair int `json:"fair"` // e.g. -80
}

type GraphColors struct {
	Good       string `json:"good"`
	Fair       string `json:"fair"`
	Poor       string `json:"poor"`
	AuthFailed string `json:"auth_failed"`
}

func DefaultGraphConfig() *GraphConfig {
	return &GraphConfig{
		RSSIThresholds: RSSIThresholds{
			Good: -65,
			Fair: -80,
		},
		Colors: GraphColors{
			Good:       "#32d74b",
			Fair:       "#ffcc00",
			Poor:       "#ff453a",
			AuthFailed: "#ff453a",
		},
	}
}

File: web/src/types/graph.ts
export type GraphGroup = 'ap' | 'station' | 'network';

export interface RSSIThresholds {
  good: number;
  fair: number;
}

export interface GraphColors {
  good: string;
  fair: string;
  poor: string;
  auth_failed: string;
}

export interface GraphConfig {
  rssi_thresholds: RSSIThresholds;
  colors: GraphColors;
}

export interface NodeIdentity {
  id: string;
  label: string;
  group: GraphGroup;
  mac?: string;
  vendor?: string;
  signature?: string;
  model?: string;
  os?: string;
  first_seen?: string; // ISO DateTime
  last_seen?: string; // ISO DateTime
}

export interface RSNInfo {
  // TODO: define fields when needed in the frontend
  [key: string]: unknown;
}

export interface WPSDetails {
  // TODO: define fields when needed in the frontend
  [key: string]: unknown;
}

export interface MobilityDomain {
  // TODO: define fields when needed in the frontend
  [key: string]: unknown;
}

export interface BSSLoad {
  station_count: number;
  channel_utilization: number; // 0-255 → 0-100%
  available_admission: number;
}

export interface RadioDetails {
  ssid?: string;
  security?: string;
  crypto?: string;
  standard?: string;
  wps_info?: string;
  handshake_file?: string;
  last_anonce?: string;
  capabilities?: string[];
  probedSSIDs?: string[];
  observed_ssids?: string[];
  ieTags?: number[];

  rsn_info?: RSNInfo;
  wps_details?: WPSDetails;
  mobility_domain?: MobilityDomain;
  bss_load?: BSSLoad;

  channel?: number;
  bw?: number;
  frequency?: number;
  rssi?: number;

  is_wifi6?: boolean;
  is_wifi7?: boolean;
  is_randomized?: boolean;
  has_handshake?: boolean;
  has11k?: boolean;
  has11v?: boolean;
  has11r?: boolean;
}

export interface TrafficStats {
  data_tx: number;
  data_rx: number;
  packets: number;
  retries: number;
}

export type ConnectionState = string;

export interface ConnectionDetails {
  connection_state?: ConnectionState;
  connection_target?: string;
  connection_error?: string;
}

export interface GraphNode extends NodeIdentity, RadioDetails, TrafficStats, ConnectionDetails {
  is_stale?: boolean;
}

export type EdgeType = 'connection' | 'probe' | 'correlation';

export interface GraphEdge {
  id?: string;
  from: string;
  to: string;
  dashed?: boolean;
  type?: EdgeType;
  label?: string;
  color?: string;
}

export interface GraphData {
  nodes: GraphNode[];
  edges: GraphEdge[];
  config?: GraphConfig;
}

export interface WSGraphMessage {
  type: 'graph';
  payload: GraphData;
}

export function isWSGraphMessage(msg: unknown): msg is WSGraphMessage {
  return (
    typeof msg === 'object' &&
    msg !== null &&
    (msg as WSGraphMessage).type === 'graph' &&
    typeof (msg as WSGraphMessage).payload === 'object'
  );
}

export type AlertSeverity = 'info' | 'warning' | 'error' | 'critical';
export type AlertType = 'anomaly' | 'ssid_match' | 'mac_match' | 'HANDSHAKE_CAPTURED' | 'ANOMALY';

export type AlertSubtype =
  | 'DEAUTH_DETECTED'
  | 'BROADCAST_DEAUTH'
  | 'HIGH_RETRY_RATE'
  | 'KARMA_DETECTION'
  | 'EVIL_TWIN_DETECTED'
  | 'OUI_SPOOFING'
  | 'KARMA_AP_DETECTED'
  | 'RULE_MATCH'
  | 'WPA_HANDSHAKE'
  | 'VULNERABILITY_DETECTED'
  | 'WEAK_CRYPTO_ZERO_NONCE'
  | 'WEAK_CRYPTO_BAD_RNG'
  | (string & {});

export interface Alert {
  id: string;
  type: AlertType;
  message: string;
  // Backend sends capitalized ("Low","Medium","High","Critical") — normalize on ingest
  severity: string;
  timestamp: string;
  device_mac?: string;
  subtype?: AlertSubtype;
  target_mac?: string;
  details?: string;
  rule_id?: string;
  handshake_file?: string;
}

export interface WSAlertMessage {
  type: 'alert';
  payload: Alert;
}

export interface WSLogMessage {
  type: 'log';
  payload: { message: string; level: string };
}

export function isWSAlertMessage(msg: unknown): msg is WSAlertMessage {
  return (
    typeof msg === 'object' &&
    msg !== null &&
    (msg as WSAlertMessage).type === 'alert' &&
    typeof (msg as WSAlertMessage).payload === 'object'
  );
}

export function isWSLogMessage(msg: unknown): msg is WSLogMessage {
  return (
    typeof msg === 'object' &&
    msg !== null &&
    (msg as WSLogMessage).type === 'log' &&
    typeof (msg as WSLogMessage).payload === 'object'
  );
}
